torch/csrc/jit/passes/reconstruct_scopes.cpp

```diff
@@ -1,83 +1,127 @@
 #include <torch/csrc/jit/passes/reconstruct_scopes.h>
 #include <torch/csrc/jit/jit_log.h>
 
+#include <iostream>
+
 namespace torch {
 namespace jit {
 
 class ReconstructScopesPass {
  public:
   ReconstructScopesPass(const Module& m, Graph& g, std::string p)
-      : root_module(&m), graph(&g), prefix(std::move(p)) {};
+      : root_module(&m), graph(&g), prefix(std::move(p)), has_duplicated_function(false) {};
   void run();
 
  private:
   const Module* root_module;
   Graph* graph;
   std::string prefix;
+  bool has_duplicated_function;
 
   std::unordered_map<Function*, ModulePtr> func_to_module;
   std::unordered_map<ModulePtr, std::string> module_names;
 
-  void visitBlock(Block* b);
-  void visitNode(Node* n);
+  void visitBlock(Block* b, const std::string& root_scope_string);
+  void visitNode(Node* n, const std::string& root_scope_string);
 
+  std::string getSubModuleName(const Module& module, const std::string& prefix);
   void constructFunctionToModuleMap(const Module& module);
   void constructRelativeNamesForModules(
       const Module& module,
       const std::string& prefix);
 
-  std::string getScopeString(Function* f) const;
+  std::string getScopeString(const InlinedCallStackEntry& frame) const;
 };
 
 void ReconstructScopesPass::constructFunctionToModuleMap(const Module& module) {
   for (auto& method : module.get_methods()) {
-    func_to_module[&method.function()] = module._ivalue();
+    Function* func_ptr = &method.function();
+    if (!has_duplicated_function &&
+        func_to_module.find(func_ptr) != func_to_module.end()) {
+      has_duplicated_function = true;
+    }
+    func_to_module[func_ptr] = module._ivalue();
   }
   for (Module m : module.children()) {
     constructFunctionToModuleMap(m);
   }
 }
 
+std::string ReconstructScopesPass::getSubModuleName(
+    const Module& module,
+    const std::string& prefix) {
+  std::string moduleType = module.type()->str();
+  size_t lastDotIndex = moduleType.rfind('.');
+  if (lastDotIndex != std::string::npos) {
+    moduleType = moduleType.substr(lastDotIndex + 1);
+  }
+  return prefix + "(" + moduleType + ")";
+}
+
 void ReconstructScopesPass::constructRelativeNamesForModules(
     const Module& module,
     const std::string& prefix) {
-  module_names[module._ivalue()] = prefix;
+  module_names[module._ivalue()] = getSubModuleName(module, prefix);
   for (NameModule s : module.named_children()) {
-    constructRelativeNamesForModules(s.value, prefix + "." + s.name);
+    std::string newPrefix = (!has_duplicated_function) ?
+        module_names[module._ivalue()] + "." + s.name : module_names[module._ivalue()] + ".";
```

Comment:
When has_duplicated_function=true, I suggest add something like <instance-name-unknown>.

Reply:
Can you clarify your suggestion a bit? Also, since we decided not to save module info the two cases differently, I believe that we don't need to check for `has_duplicated_function` here. Is that correct?

Reply:
> Can you clarify your suggestion a bit? Also, since we decided not to save module info the two cases differently, I believe that we don't need to check for `has_duplicated_function` here. Is that correct?

Are you talking about not checking `has_duplicated_function` for this PR or after the other PR

Reply:
> When has_duplicated_function=true, I suggest add something like .

My first comment had trail off. I meant when has_duplicated_function=true, add ".<instance info unavailable>" or something like that.